plugins/Oscilloscope/OscilloscopeGraph.cpp

```diff
@@ -0,0 +1,168 @@
+*
+* This program is free software; you can redistribute it and/or
+* modify it under the terms of the GNU General Public
+* License as published by the Free Software Foundation; either
+* version 2 of the License, or (at your option) any later version.
+*
+* This program is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+* General Public License for more details.
+*
+* You should have received a copy of the GNU General Public
+* License along with this program (see COPYING); if not, write to the
+* Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+* Boston, MA 02110-1301 USA.
+*
+*/
+
+#include "OscilloscopeGraph.h"
+#include "OscilloscopeControls.h"
+#include "Oscilloscope.h"
+#include "embed.h"
+#include "GuiApplication.h"
+#include "MainWindow.h"
+
+#include <QPainter>
+#include <QSizePolicy>
+#include <QWheelEvent>
+
+#include <functional>
+
+namespace lmms::gui
+{
+
+OscilloscopeGraph::OscilloscopeGraph(QWidget* parent, OscilloscopeControls* controls):
+	QWidget(parent),
+	m_controls(controls),
+	m_inputBufferReader(static_cast<Oscilloscope*>(m_controls->effect())->inputBuffer())
+{
+	setAutoFillBackground(true);
+	setMinimumSize(400, 200);
+	setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Expanding);
+
+	connect(getGUI()->mainWindow(), SIGNAL(periodicUpdate()), this, SLOT(update()));
+	connect(Engine::audioEngine(), &AudioEngine::sampleRateChanged, this, &OscilloscopeGraph::changeSampleRate);
+	changeSampleRate();
+}
+
+void OscilloscopeGraph::changeSampleRate()
+{
+	int newBufferSize = Engine::audioEngine()->outputSampleRate() * MaxBufferLengthSeconds;
+	m_ringBuffer.resize(newBufferSize);
+	m_ringBufferIndex = m_ringBufferIndex % newBufferSize;
+}
+
+void OscilloscopeGraph::paintEvent(QPaintEvent* pe)
+{
+	Oscilloscope* effect = static_cast<Oscilloscope*>(m_controls->effect());
+
+	// Update the ring buffer with any new data from the audio thread
+	auto incomingBuffer = m_inputBufferReader.read_max(effect->inputBuffer().capacity());
+	for (f_cnt_t f = 0; f < incomingBuffer.size(); ++f)
+	{
+		m_ringBuffer[m_ringBufferIndex] = incomingBuffer[f];
+		m_ringBufferIndex = (m_ringBufferIndex + 1) % m_ringBuffer.size();
+    }
+
+	QPainter p(this);
+	p.fillRect(0, 0, width(), height(), m_backgroundColor);
+
+	float amp = m_controls->m_ampModel.value() * 0.01f;
+	float phase = m_controls->m_phaseModel.value();
+	int windowSizeFrames = m_ringBuffer.size() * m_controls->m_lengthModel.value() / m_controls->m_lengthModel.maxValue();
+	int framesPerPixel = std::max(1, windowSizeFrames / width());
+
+	int windowStartIndex = m_ringBufferIndex + (m_ringBuffer.size() - windowSizeFrames) + (phase) * m_ringBuffer.size();
+	// Quantize start index to prevent flickering (round up to prevent going below the current ring buffer index)
+	int frameOffsetWithinPixel = windowStartIndex % framesPerPixel;
```

Comment:
(Non-blocking)

- `amp`, `framesPerPixel`, and `frameOffsetWithinPixel` are never mutated and should be declared `const`
- `phase` is used exactly once and can be inlined (it's clear that it's the phase even without this variable binding)
- Some of these lines exceed 120 characters in length, breaking them on operators resolves this

```suggestion
	const float amp = m_controls->m_ampModel.value() * 0.01f;
	int windowSizeFrames = m_ringBuffer.size()
		* m_controls->m_lengthModel.value()
		/ m_controls->m_lengthModel.maxValue();
	const int framesPerPixel = std::max(1, windowSizeFrames / width());

	int windowStartIndex = m_ringBufferIndex
		+ (m_ringBuffer.size() - windowSizeFrames)
		+ m_controls->m_phaseModel.value() * m_ringBuffer.size();
	// Quantize start index to prevent flickering (round up to prevent going below the current ring buffer index)
	const int frameOffsetWithinPixel = windowStartIndex % framesPerPixel;
```